Sources/Sentry/SentryMetricProfiler.mm

```diff
@@ -0,0 +1,156 @@
+
+- (void)start
+{
+    [self registerTimeSeriesHandler];
+    [self registerMemoryPressureWarningHandler];
+}
+
+- (void)stop
+{
+    [_timer invalidate];
+    dispatch_source_cancel(_memoryWarningSource);
+    [_notificationCenter removeObserver:self name:NSProcessInfoThermalStateDidChangeNotification];
+}
+
+- (NSData *)serialize
+{
+    // TODO: implement
+    return [[NSData alloc] init];
+}
+
+#pragma mark - Private
+
+- (void)registerTimeSeriesHandler
+{
+    _timer = [NSTimer scheduledTimerWithTimeInterval:kSentryMetricProfilerInterval
+                                             repeats:YES
+                                               block:^(NSTimer *_Nonnull timer) {
+                                                   [self recordCPUPercentage];
+                                                   [self recordMemoryFootprint];
+                                               }];
+}
+
+/**
+ * This is a more fine-grained API, providing normal/warn/critical levels of memory usage, versus
+ * using `UIApplicationDidReceiveMemoryWarningNotification` which does not provide any additional
+ * information ("This notification does not contain a userInfo dictionary." from
+ * https://developer.apple.com/documentation/uikit/uiapplication/1622920-didreceivememorywarningnotificat).
+ */
+- (void)registerMemoryPressureWarningHandler
+{
+    const auto queueAttributes
+        = dispatch_queue_attr_make_with_qos_class(DISPATCH_QUEUE_SERIAL, QOS_CLASS_UTILITY, 0);
+    _memoryWarningQueue = dispatch_queue_create("io.sentry.queue.memory-warnings", queueAttributes);
+    _memoryWarningSource = dispatch_source_create(DISPATCH_SOURCE_TYPE_MEMORYPRESSURE, 0,
+        DISPATCH_MEMORYPRESSURE_NORMAL | DISPATCH_MEMORYPRESSURE_WARN
+            | DISPATCH_MEMORYPRESSURE_CRITICAL,
+        _memoryWarningQueue);
+    dispatch_source_set_event_handler(_memoryWarningSource, ^{
+        [self recordMemoryPressureState:dispatch_source_get_data(self->_memoryWarningSource)];
+    });
+    dispatch_resume(_memoryWarningSource);
+}
+
+- (void)registerStateChangeNotifications
+{
+    // According to Apple docs: "To receive NSProcessInfoThermalStateDidChangeNotification, you must
+    // access the thermalState prior to registering for the notification." (from
+    // https://developer.apple.com/documentation/foundation/nsprocessinfothermalstatedidchangenotification/)
+    [self recordThermalState];
+
+    // According to Apple docs: "This notification is posted on the global dispatch queue."
+    [_notificationCenter addObserver:self
+                            selector:@selector(handleThermalStateChangeNotification:)
+                                name:NSProcessInfoThermalStateDidChangeNotification];
+}
+
+- (void)handleThermalStateChangeNotification:(NSNotification *)note
+{
+    [self recordThermalState];
+}
+
+- (void)recordThermalState
+{
+    [_thermalStateChanges
+        addObject:[self metricEntryForValue:@(NSProcessInfo.processInfo.thermalState)]];
```

Comment:
`m`: We could create a wrapper around `NSProcessInfo` for testing.

Reply:
I added two new wrappers to help test this implementation: 
- [`SentryNSProcessInfoWrapper`](https://github.com/getsentry/sentry-cocoa/pull/2493/files#diff-2fab6e78e2dace206d0e3db798d99176d8e2c12d41fcfed70da04ef404b4bd36)
    - This will also handle the specific notifications we observe, but they do so via `SentryNSNotificationCenterWrapper` so all will still be testable
- [`SentrySystemWrapper`](https://github.com/getsentry/sentry-cocoa/pull/2493/files#diff-9e1f34f1708681d0399c883a1b551764116d23f3d081f362b487dbc703f28f53)
    - This is a good place for us to mock out any calls to `sysctl`, `mach` etc APIs